File: lib/providers/octopart.ts
import { ProviderResult } from "@/types/bom";
import {
  PartProvider,
  OctopartSearchResponse,
} from "./types";

const NEXAR_TOKEN_URL = "https://identity.nexar.com/connect/token";
const NEXAR_API_URL = "https://api.nexar.com/graphql";

// Token cache for OAuth
let cachedToken: { token: string; expiresAt: number } | null = null;

export class OctopartProvider implements PartProvider {
  name = "octopart";
  private clientId: string;
  private clientSecret: string;

  constructor() {
    this.clientId = process.env.OCTOPART_CLIENT_ID || "";
    this.clientSecret = process.env.OCTOPART_CLIENT_SECRET || "";
  }

  private async getAccessToken(): Promise<string> {
    // Check if we have a valid cached token (with 60s buffer)
    if (cachedToken && cachedToken.expiresAt > Date.now() + 60000) {
      return cachedToken.token;
    }

    if (!this.clientId || !this.clientSecret) {
      throw new Error("Octopart OAuth credentials not configured");
    }

    const response = await fetch(NEXAR_TOKEN_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: new URLSearchParams({
        grant_type: "client_credentials",
        client_id: this.clientId,
        client_secret: this.clientSecret,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`OAuth token request failed: ${response.status} - ${errorText}`);
    }

    const data = await response.json();

    // Cache the token
    cachedToken = {
      token: data.access_token,
      expiresAt: Date.now() + (data.expires_in * 1000),
    };

    console.log("Obtained new Octopart OAuth token");
    return cachedToken.token;
  }

  async search(query: string): Promise<ProviderResult[]> {
    return this.executeSearch(query);
  }

  async searchByMPN(mpn: string, manufacturer?: string): Promise<ProviderResult[]> {
    const query = manufacturer ? `${manufacturer} ${mpn}` : mpn;
    return this.executeSearch(query);
  }

  private async executeSearch(query: string): Promise<ProviderResult[]> {
    if (!this.clientId || !this.clientSecret) {
      console.warn("No Octopart OAuth credentials configured, returning mock data");
      return this.getMockResults(query);
    }

    const graphqlQuery = `
      query Search($q: String!) {
        supSearch(q: $q, limit: 10) {
          results {
            part {
              mpn
              manufacturer {
                name
              }
              shortDescription
              sellers {
                company {
                  name
                }
                offers {
                  clickUrl
                  inventoryLevel
                  moq
                  prices {
                    price
                    currency
                    quantity
                  }
                }
              }
            }
          }
        }
      }
    `;

    try {
      const token = await this.getAccessToken();

      const response = await fetch(NEXAR_API_URL, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({
          query: graphqlQuery,
          variables: { q: query },
        }),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Nexar API error: ${response.status} - ${errorText}`);
      }

      const data = await response.json();

      if (data.errors) {
        console.error("GraphQL errors:", data.errors);
        throw new Error(data.errors[0]?.message || "GraphQL error");
      }

      return this.transformResults(data);
    } catch (error) {
      console.error("Octopart search failed:", error);
      return this.getMockResults(query);
    }
  }

  private transformResults(data: { data: { supSearch: { results: Array<{ part: {
    mpn: string;
    manufacturer: { name: string };
    shortDescription: string;
    sellers: Array<{
      company: { name: string };
      offers: Array<{
        clickUrl: string;
        inventoryLevel: number;
        moq: number | null;
        prices: Array<{ price: number; currency: string; quantity: number }>;
      }>;
    }>;
  } }> } } }): ProviderResult[] {
    const results: ProviderResult[] = [];

    if (!data.data?.supSearch?.results) {
      return results;
    }

    for (const result of data.data.supSearch.results) {
      const part = result.part;
      if (!part.sellers) continue;

      for (const seller of part.sellers) {
        for (const offer of seller.offers) {
          if (!offer.prices || offer.prices.length === 0) continue;

          // Get the lowest price tier
          const lowestPrice = offer.prices.reduce((min, p) =>
            p.price < min.price ? p : min
          );

          results.push({
            partNumber: part.mpn,
            manufacturer: part.manufacturer?.name || "Unknown",
            description: part.shortDescription || "",
            price: lowestPrice.price,
            currency: lowestPrice.currency,
            stock: offer.inventoryLevel || 0,
            minQuantity: offer.moq || 1,
            provider: this.name,
            distributor: seller.company?.name || "Unknown",
            url: offer.clickUrl || "",
          });
        }
      }
    }

    // Sort by price
    return results.sort((a, b) => a.price - b.price);
  }

  private getMockResults(query: string): ProviderResult[] {
    // Return mock data when credentials are not configured
    return [
      {
        partNumber: query.toUpperCase(),
        manufacturer: "Example Mfg",
        description: `Mock result for "${query}"`,
        price: 1.25,
        currency: "USD",
        stock: 1000,
        minQuantity: 1,
        provider: this.name,
        distributor: "DigiKey",
        url: "https://www.digikey.com",
      },
      {
        partNumber: query.toUpperCase(),
        manufacturer: "Example Mfg",
        description: `Mock result for "${query}"`,
        price: 1.15,
        currency: "USD",
        stock: 500,
        minQuantity: 10,
        provider: this.name,
        distributor: "Mouser",
        url: "https://www.mouser.com",
      },
      {
        partNumber: query.toUpperCase(),
        manufacturer: "Example Mfg",
        description: `Mock result for "${query}"`,
        price: 0.95,
        currency: "USD",
        stock: 2500,
        minQuantity: 100,
        provider: this.name,
        distributor: "Arrow",
        url: "https://www.arrow.com",
      },
    ];
  }
}

File: python-agent/src/bom_agent_service/providers/octopart.py
"""Octopart/Nexar part provider with OAuth 2.0 and GraphQL."""

import logging
import os
import time
from typing import Optional

import httpx

from .types import PartProvider, ProviderResult

logger = logging.getLogger(__name__)

NEXAR_TOKEN_URL = "https://identity.nexar.com/connect/token"
NEXAR_API_URL = "https://api.nexar.com/graphql"

SEARCH_QUERY = """
query Search($q: String!) {
  supSearch(q: $q, limit: 10) {
    results {
      part {
        mpn
        manufacturer {
          name
        }
        shortDescription
        sellers {
          company {
            name
          }
          offers {
            clickUrl
            inventoryLevel
            moq
            prices {
              price
              currency
              quantity
            }
          }
        }
      }
    }
  }
}
"""


class TokenCache:
    """In-memory OAuth token cache."""

    def __init__(self):
        self.token: Optional[str] = None
        self.expires_at: float = 0


_token_cache = TokenCache()


class OctopartProvider(PartProvider):
    """Octopart/Nexar part provider using OAuth 2.0 and GraphQL."""

    name = "octopart"

    def __init__(self):
        self.client_id = os.getenv("OCTOPART_CLIENT_ID", "")
        self.client_secret = os.getenv("OCTOPART_CLIENT_SECRET", "")

    def is_configured(self) -> bool:
        """Check if Octopart credentials are configured."""
        return bool(self.client_id and self.client_secret)

    async def _get_access_token(self) -> str:
        """Get OAuth access token with caching."""
        # Check if we have a valid cached token (with 60s buffer)
        if _token_cache.token and _token_cache.expires_at > time.time() + 60:
            return _token_cache.token

        if not self.client_id or not self.client_secret:
            raise ValueError("Octopart OAuth credentials not configured")

        async with httpx.AsyncClient() as client:
            response = await client.post(
                NEXAR_TOKEN_URL,
                data={
                    "grant_type": "client_credentials",
                    "client_id": self.client_id,
                    "client_secret": self.client_secret,
                },
                headers={"Content-Type": "application/x-www-form-urlencoded"},
            )
            response.raise_for_status()
            data = response.json()

            _token_cache.token = data["access_token"]
            _token_cache.expires_at = time.time() + data["expires_in"]

            logger.info("Obtained new Octopart OAuth token")
            return _token_cache.token

    async def search(self, query: str) -> list[ProviderResult]:
        """Search for parts by query."""
        return await self._execute_search(query)

    async def search_by_mpn(
        self, mpn: str, manufacturer: Optional[str] = None
    ) -> list[ProviderResult]:
        """Search for parts by MPN."""
        query = f"{manufacturer} {mpn}" if manufacturer else mpn
        return await self._execute_search(query)

    async def _execute_search(self, query: str) -> list[ProviderResult]:
        """Execute GraphQL search against Nexar API."""
        if not self.is_configured():
            raise ValueError(
                "Octopart credentials not configured. "
                "Set OCTOPART_CLIENT_ID and OCTOPART_CLIENT_SECRET."
            )

        token = await self._get_access_token()

        async with httpx.AsyncClient() as client:
            response = await client.post(
                NEXAR_API_URL,
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                },
                json={
                    "query": SEARCH_QUERY,
                    "variables": {"q": query},
                },
                timeout=30.0,
            )
            response.raise_for_status()
            data = response.json()

            # Check for GraphQL errors
            if "errors" in data:
                error_msg = data["errors"][0].get("message", "Unknown GraphQL error")
                logger.error(f"GraphQL errors: {data['errors']}")
                raise ValueError(f"Nexar API error: {error_msg}")

            return self._transform_results(data)

    def _transform_results(self, data: dict) -> list[ProviderResult]:
        """Transform Nexar GraphQL response to ProviderResult list."""
        results: list[ProviderResult] = []

        sup_search = data.get("data", {}).get("supSearch", {})
        search_results = sup_search.get("results", [])
        if not search_results:
            return results

        for result in search_results:
            part = result.get("part", {})
            sellers = part.get("sellers", [])
            if not sellers:
                continue

            mpn = part.get("mpn", "")
            manufacturer = part.get("manufacturer", {}).get("name", "Unknown")
            description = part.get("shortDescription", "")

            for seller in sellers:
                company_name = seller.get("company", {}).get("name", "Unknown")
                offers = seller.get("offers", [])

                for offer in offers:
                    prices = offer.get("prices", [])
                    if not prices:
                        continue

                    # Find the lowest price
                    lowest_price = min(prices, key=lambda p: p.get("price", float("inf")))
                    price = lowest_price.get("price", 0)
                    if price <= 0:
                        continue

                    moq = offer.get("moq") or 1
                    inventory = offer.get("inventoryLevel") or 0

                    results.append(
                        ProviderResult(
                            mpn=mpn,
                            manufacturer=manufacturer,
                            description=description,
                            price=price,
                            currency=lowest_price.get("currency", "USD"),
                            stock=inventory,
                            min_quantity=moq,
                            provider=self.name,
                            distributor=company_name,
                            url=offer.get("clickUrl", ""),
                        )
                    )

        return sorted(results, key=lambda r: r.price)
